platform/view/core/endpoint/endpoint.go

```diff
@@ -241,20 +236,22 @@ func (r *Service) SetPublicKeyIDSynthesizer(publicKeyIDSynthesizer driver.Public
 }
 
 func (r *Service) pkiResolve(resolver *Resolver) []byte {
-	r.pkiExtractorsLock.RLock()
+	resolver.PKILock.RLock()
 	if len(resolver.PKI) != 0 {
-		r.pkiExtractorsLock.RUnlock()
 		return resolver.PKI
 	}
-	r.pkiExtractorsLock.RUnlock()
+	resolver.PKILock.RUnlock()
 
+	resolver.PKILock.Lock()
+	defer resolver.PKILock.Unlock()
 	resolver.PKI = r.ExtractPKI(resolver.Id)
```

Comment:
check again if resolver.PKI is different from nil. Another thread might have already set it.